functional/connectivity-loss_test.go

```diff
@@ -0,0 +1,229 @@
+	slack := 2 * time.Second
+
+	time.Sleep(ttl + agentReconcileInterval + slack)
+
+	// Check unit state after connection loss.
+	//
+	// Note: we cannot use fleetctl to check the state here,
+	// as fleet is not available to give us this information...
+	// We have to go deeper, and try to obtain the information from systemd directly.
+	stdout, err := cluster.MemberCommand(m0, "sudo", "sh", "-c", `'systemctl list-unit-files -t service --no-legend | ( grep "\(single\|global\)@" || true )'`)
+	stdout = strings.TrimSpace(stdout)
+	if err != nil {
+		t.Fatalf("Failed to retrieve systemd unit file states: %v", err)
+	}
+	actualSystemdFiles := map[string]string{}
+	if stdout != "" {
+		for _, line := range strings.Split(stdout, "\n") {
+			fields := strings.Fields(line)
+			actualSystemdFiles[fields[0]] = fields[1]
+		}
+	}
+	if !reflect.DeepEqual(actualSystemdFiles, expectedSystemdFiles) {
+		t.Fatalf("Units files not in expected state after loosing connectivity.\nExpected: %v\nActual: %v", expectedSystemdFiles, actualSystemdFiles)
+	}
+
+	stdout, err = cluster.MemberCommand(m0, "sudo", "sh", "-c", `'systemctl list-units -t service --no-legend | ( grep "\(single\|global\)@" || true )'`)
+	stdout = strings.TrimSpace(stdout)
+	if err != nil {
+		t.Fatalf("Failed to retrieve systemd unit states: %v", err)
+	}
+	actualSystemdStates := map[string][]string{}
+	if stdout != "" {
+		for _, line := range strings.Split(stdout, "\n") {
+			fields := strings.Fields(line)
+			actualSystemdStates[fields[0]] = fields[1:4]
+		}
+	}
+	if !reflect.DeepEqual(actualSystemdStates, expectedSystemdStates) {
+		t.Fatalf("Units not in expected state after loosing connectivity.\nExpected: %v\nActual: %v", expectedSystemdStates, actualSystemdStates)
+	}
+
+	// Restore etcd connection.
+	if _, err = cluster.MemberCommand(m0, "sudo", "iptables", "-D", "OUTPUT", "-p", "tcp", "-m", "multiport", "--dports=2379,4001", "-j", "REJECT"); err != nil {
+		t.Fatal(err)
+	}
+
+	// Again, wait long enough to be reasonably confident that no more state changes will happen.
+	//
+	// Here this should cover the time for fleet to realise connectivity is back,
+	// and for the Agent to complete the second run after reconnection.
+	//
+	// (Unlike for the first run immediately after connectivity is back, by the time of the second run,
+	// Engine leadership and Engine reconciliation should have been sorted out,
+	// and everything should be back to normal...)
+	time.Sleep(ttl + agentReconcileInterval + slack)
+
+	// Check state after reconnect.
+	if isExpected, expected, actual := checkExpectedStates(); !isExpected {
+		t.Fatalf("Units not in expected state after restoring connectivity.\nExpected: %v\nActual: %v", expected, actual)
+	}
+
+	// Additionally check the logs of all active units for possible temporary state flapping.
+	stdout, err = cluster.MemberCommand(m0, "sudo", "sh", "-c", `journalctl _PID=1 | ( grep 'Stopping \(single\|global\)@' || true )`)
```

Comment:
Yes, I could do the filtering in Go; but it would be more complicated, for no real gain as far as I can tell -- so I decided against it.